WalletWasabi.Tests/UnitTests/Crypto/ProofSystemTests.cs

```diff
@@ -40,6 +42,53 @@ public void CanProveAndVerifyMAC()
 			// was generated with the coordinator's secret key.
 			var clientStatement = ProofSystem.IssuerParameters(coordinatorParameters, mac, Ma);
 			var isValidProof = ProofSystem.Verify(clientStatement, proofOfMac);
+			Assert.True(isValidProof);
+
+			var corruptedResponses = new ScalarVector(proofOfMac.Responses.Reverse());
+			var invalidProofOfMac = new Proof(proofOfMac.PublicNonces, corruptedResponses);
+			isValidProof = ProofSystem.Verify(clientStatement, invalidProofOfMac);
+			Assert.False(isValidProof);
+
+			var corruptedPublicNonces = new GroupElementVector(proofOfMac.PublicNonces.Reverse());
+			invalidProofOfMac = new Proof(corruptedPublicNonces, proofOfMac.Responses);
+			isValidProof = ProofSystem.Verify(clientStatement, invalidProofOfMac);
+			Assert.False(isValidProof);
+		}
+
+		[Fact]
+		[Trait("UnitTest", "UnitTest")]
+		public void CanProveAndVerifyMacShow()
+		{
+			var rnd = new SecureRandom();
+			var coordinatorKey = new CoordinatorSecretKey(rnd);
+			var coordinatorParameters = coordinatorKey.ComputeCoordinatorParameters();
+
+			// A blinded amount is known as an `attribute`. In this case the attribute Ma is the
+			// value 10000 blinded with a random `blindingFactor`. This attribute is sent to
+			// the coordinator.
+			var amount = new Scalar(10_000);
+			var r = rnd.GetScalar();
+			var Ma = amount * Generators.Gg + r * Generators.Gh;
+
+			// The coordinator generates a MAC and a proof that the MAC was generated using the
+			// coordinator's secret key. The coordinator sends the pair (MAC, proofOfMac) back
+			// to the client.
+			var t = rnd.GetScalar();
+			var mac = MAC.ComputeMAC(coordinatorKey, Ma, t);
+
+			// The client randomizes the commitments before presenting them to the coordinator proving to
+			// the coordinator that a credential is valid (prover knows a valid MAC on non-randomized attribute)
+			var z = rnd.GetScalar();
+			var credential = CredentialPresentation.FromMAC(mac, z, amount, r);
+			var knowledge = ProofSystem.ShowCredential(credential, z, mac.T, amount, r, coordinatorParameters);
```

Comment:
What do you think about:

```cs
class Credential {
    public Scalar Amount { get; } // could consider uint, but only really makes sense for credential requests because of range proofs
    public Scalar Randomness { get; }
    public MAC Mac { get; }

    public (CredentialPresentation Presentation, Knowledge ShowKnowledge) Present(CoordinatorParameters iparams, Scalar z)
        => CredentialPresentation.FromMAC(Mac, z, Amount, Randomness); // or just merge into here
}
```

```cs
class ProofSystem
{
    public Knowledge ShowCredential(CoordinatorParameters, CredentialPresentation presentation, Credential credential) // note argument order reversed, it's customary to put public inputs before witnesses in most PoK notation

    public Statement ShowCredential(CoordinatorParameters, Z, CredentialPresentation presentation)
}
```

Reply:
It is a good idea. In my mind a `Credential` was what we send to the coordinator but we can call credential to this and call `CredentialRequest` to the tuple (Ma, rangeProof), `CredentialPresentation` to the tuple (blindedCredential, showCredentialProof)  and finally BlindedCredential to the randomized commitments.

What we currently call `CredentailPresentation` should be what we send to the coordinator.